hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/commit/UpsertPartitioner.java

```diff
@@ -158,13 +162,17 @@ private List<SmallFile> filterSmallFilesInClustering(final Set<String> pendingCl
   private void assignInserts(WorkloadProfile profile, HoodieEngineContext context) {
     // for new inserts, compute buckets depending on how many records we have for each partition
     Set<String> partitionPaths = profile.getPartitionPaths();
-    long averageRecordSize =
-        averageBytesPerRecord(table.getMetaClient().getActiveTimeline().getCommitTimeline().filterCompletedInstants(),
-            config);
+    /*
+     * NOTE: we only use commit instants to calculate average record size because replacecommit can be
+     * created by clustering, which has smaller average record size, which affects assigning inserts and
+     * may result in OOM by making spark underestimate the actual input record sizes.
+     */
+    long averageRecordSize = averageBytesPerRecord(table.getMetaClient().getActiveTimeline()
+        .getTimelineOfActions(CollectionUtils.createSet(COMMIT_ACTION)).filterCompletedInstants(), config);
```

Comment:
The change looks good. Is it possible to reproduce a workload for this kind of scenario?

Reply:
simulated the scenario: COW inserts created small files, and then cluster created larger files.

Clustered file

```
4.4M Oct  5 18:33 9b3b96be-95ca-4e93-9a08-ab49640bd113-0_2-99-376_20221005183355230.parquet

// cat .hoodie/xxx.replacecommit
      "numWrites" : 39781,
      "totalWriteBytes" : 4664174,
// avg ≈ 117
```

Inserted file

```
863K Oct  5 18:33 d88007a8-f5d9-4c97-a5f1-56907b99d4a1-0_23-77-286_20221005183349960.parquet

// cat .hoodie/xxx.commit
      "numWrites" : 3588,
      "totalWriteBytes" : 818456,
// avg ≈ 228
```

Reply:
Cool. Thanks for sharing.